lib/lambda_ethereum_consensus/state_transition/epoch_processing.ex

```diff
@@ -344,8 +345,205 @@ defmodule LambdaEthereumConsensus.StateTransition.EpochProcessing do
+        {:error, reason} -> {:error, reason}
+      end
+    end
+  end
+
+  defp weigh_justification_and_finalization(
+         state,
+         total_active_balance,
+         previous_epoch_target_balance,
+         current_epoch_target_balance
+       ) do
+    previous_epoch = Accessors.get_previous_epoch(state)
+    current_epoch = Accessors.get_current_epoch(state)
+    old_previous_justified_checkpoint = state.previous_justified_checkpoint
+    old_current_justified_checkpoint = state.current_justified_checkpoint
+
+    with {:ok, previous_block_root} <- Accessors.get_block_root(state, previous_epoch),
+         {:ok, current_block_root} <- Accessors.get_block_root(state, current_epoch) do
+      new_state =
+        state
+        |> update_first_bit()
+        |> update_previous_epoch_justified(
+          previous_epoch_target_balance * 3 >= total_active_balance * 2,
+          previous_epoch,
+          previous_block_root
+        )
+        |> update_current_epoch_justified(
+          current_epoch_target_balance * 3 >= total_active_balance * 2,
+          current_epoch,
+          current_block_root
+        )
+        |> update_checkpoint_finalization_1(old_previous_justified_checkpoint, current_epoch)
+        |> update_checkpoint_finalization_2(old_previous_justified_checkpoint, current_epoch)
+        |> update_checkpoint_finalization_3(old_current_justified_checkpoint, current_epoch)
+        |> update_checkpoint_finalization_4(old_current_justified_checkpoint, current_epoch)
+
+      {:ok, new_state}
+    else
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  defp update_first_bit(state) do
+    bits =
+      state.justification_bits
+      |> BitVector.new(4)
+      |> BitVector.shift_higher(1)
+      |> BitVector.to_byte()
+
+    %BeaconState{
+      state
+      | previous_justified_checkpoint: state.current_justified_checkpoint,
+        justification_bits: bits
+    }
+  end
+
+  @spec update_previous_epoch_justified(
+          BeaconState.t(),
+          boolean,
+          SszTypes.epoch(),
+          SszTypes.root()
+        ) :: BeaconState.t()
+  defp update_previous_epoch_justified(state, is_true, previous_epoch, previous_block_root) do
+    if !is_true do
+      state
+    else
+      new_checkpoint = %SszTypes.Checkpoint{
+        epoch: previous_epoch,
+        root: previous_block_root
+      }
+
+      bits =
+        state.justification_bits
+        |> BitVector.new(4)
+        |> BitVector.set(1)
+        |> BitVector.to_byte()
+
+      %BeaconState{
+        state
+        | current_justified_checkpoint: new_checkpoint,
+          justification_bits: bits
+      }
+    end
+  end
+
+  @spec update_current_epoch_justified(
+          BeaconState.t(),
+          boolean,
+          SszTypes.epoch(),
+          SszTypes.root()
+        ) :: BeaconState.t()
+  defp update_current_epoch_justified(state, is_true, current_epoch, current_block_root) do
+    if !is_true do
+      state
+    else
```

Comment:
This can be done by matching in the function signature:

```elixir
defp update_current_epoch_justified(state, true, current_epoch, current_block_root) do
  ...
end

defp update_current_epoch_justified(state, false, current_epoch, current_block_root) do
  ...
end